### PR TITLE
Sanitizing the IRI

### LINKED ISSUE #57: Allow for proper IRI escaping
When constructing IRI's I need to use JSONPath results that are not properly escaped and the resulting IRI wil be invalid. Sometimes it could be as simple as replacing space (' ') with  an underscore ('_'), but none of that is possible with JSONPath and the properties files configuration. 
Could we have an option to always (properly) escape the IRI's, respecting query parameters and fragments. 

Solution suggestion (by Claude):
```
import java.net.URI;
import org.apache.http.client.utils.URIBuilder;

URIBuilder builder = new URIBuilder("https://example.com/page?title=book title#a b");
URI uri = builder.build();
System.out.println(uri.toASCIIString());
// -> https://example.com/page?title=book%20title#a%20b
```
Maybe make it into an option; like `forceURIEncoding`?

### PR DESCRIPTION
**Pull request by PaulBoon**
Closes [#57](https://github.com/gdcc/exporter-dcat3/issues/57)

Firts it is replacing whitespace with underscores. 
If the result is not a valid IRI, encode all characters except for the unicode letters; keep readabilty that we have with IRI's.

### CODE REVIEW

**Comment by sjaakd** [comment]
Hi paul.. Why do you need a configuration flag for this? I would say its always wrong to encode illegal characters. So preferable I leave the flag out of the solution. Or are the some datasets in your selection that require illegal characters in IRI's ?

File: src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java [hunk 1]
```diff
@@ -0,0 +1,282 @@
+package io.gdcc.spi.export.dcat3.config.validate;
+import java.nio.charset.StandardCharsets;
+import java.util.regex.*;
+
+public class IRIFixer {
```
Comment (sjaakd) [inline src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java]: I would call it `IriSanitizer` or `IriCodec`.

File: src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+     *
+     * <ul>
+     *   <li>group 2 = scheme</li>
+     *   <li>group 4 = authority</li>
+     *   <li>group 5 = path</li>
+     *   <li>group 7 = query</li>
+     *   <li>group 9 = fragment</li>
+     * </ul>
+     */
+    private static final Pattern SPLIT = Pattern.compile(
```
Comment (sjaakd) [inline src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java]: why not use `java.net.URI parsing` first, then encode components, and keep regex only as a small fallback for edge cases.

Reason: URI parsing is clearer and less error-prone than a big RFC regex splitter, especially for authority/query/fragment nuances and maintenance over time. And your unit test can be smaller.

File: src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java [hunk 3]
```diff
@@ -0,0 +1,282 @@
+package io.gdcc.spi.export.dcat3.config.validate;
```
Comment (sjaakd) [inline src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java]: I guess it belongs in the mapping package? It's a mapping util right?

File: src/test/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixerTest.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        "'', ''"
+    })
+    // spotless:on
+    @DisplayName("Encode URI/IRI components using RFC 3986 safe characters")
+    void buildValidUri_cases(String raw, String expected) {
+        assertThat(IRIFixer.buildValidUri(raw)).isEqualTo(expected);
+    }
+
+    @ParameterizedTest(name = "buildValidIri(''{0}'') => ''{1}''")
+    // spotless:off
```
Comment (sjaakd) [inline src/test/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixerTest.java]: fighting spotless as well. I don't like it to be honest.

File: src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java (hunk 2 above)
Comment (sjaakd) [reply src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java]: there's also the possibility for gaps:

- It does not fully validate authority semantics (it treats authority mostly as a character bag, so structurally invalid host/userinfo forms can slip through).
- The Unicode check is simplified; some noncharacter code points can still be treated as valid.
- Existing %HH sequences are not preserved in encoding paths, so inputs can be double-encoded (%20 → %2520) when another illegal char triggers sanitization.

File: src/main/java/io/gdcc/spi/export/dcat3/mapping/ResourceMapper.java [hunk 5]
```diff
@@ -461,6 +475,19 @@ private static boolean looksLikeIri(String s) {
         return s != null && s.matches("^[a-zA-Z][a-zA-Z0-9+.-]*:.*");
     }
 
+    private String sanitizeIri(String iri) {
+        // Replace whitespace with underscores, takes care of most issues
+        iri = iri.replaceAll("\\s+", "_");
```
Comment (sjaakd) [inline src/main/java/io/gdcc/spi/export/dcat3/mapping/ResourceMapper.java]: why not move this to the sanitizer as well? Its the same concern.

File: src/main/java/io/gdcc/spi/export/dcat3/mapping/ResourceMapper.java [hunk 6]
```diff
@@ -461,6 +475,19 @@ private static boolean looksLikeIri(String s) {
...
+
+        // Try to fix all characters
+        // Maybe make this into an option ( or forcing into Uri)?
+        // Note that we could have the whitespace percent encoded as well, but we don't.
+        if (encodeInvalidIris && !IRIFixer.isValidIri(iri)) { // prevent double encoding
```
Comment (sjaakd) [inline src/main/java/io/gdcc/spi/export/dcat3/mapping/ResourceMapper.java]: see general comments.. I would not make this configurable.

File: src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java (hunk 2 above)
Comment (sjaakd) [reply src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java]: I took the liberty to ask copilot to use apache Jena (already on the cp)  and `java.net.URI ` to rewrite this class. Note the "build" methods have been replaced by "to". It then looks like:

```java
package io.gdcc.spi.export.dcat3.mapping.util;

import java.net.URI;
import java.net.URISyntaxException;
import java.nio.charset.StandardCharsets;
import org.apache.jena.iri.IRI;
import org.apache.jena.iri.IRIFactory;

public final class IriSanitizer {

    private IriSanitizer() {}

    private static final String SAFE_AUTH = "-._~!$&'()*+,;=:@[]";
    private static final String SAFE_PATH = "-._~!$&'()*+,;=:@/";
    private static final String SAFE_QUERY_FRAGMENT = "-._~!$&'()*+,;=:@/?";

    private static final IRIFactory IRI_FACTORY = IRIFactory.iriImplementation();

    public static String toValidIri(String raw) {
        if (raw == null) return null;
        Components c = split(raw);

        StringBuilder out = new StringBuilder();
        if (c.scheme != null) out.append(c.scheme).append(':');
        if (c.authority != null) out.append("//").append(encode(c.authority, SAFE_AUTH, true));
        if (c.path != null) out.append(encode(c.path, SAFE_PATH, true));
        if (c.query != null) out.append('?').append(encode(c.query, SAFE_QUERY_FRAGMENT, true));
        if (c.fragment != null) out.append('#').append(encode(c.fragment, SAFE_QUERY_FRAGMENT, true));
        return out.toString();
    }

    public static String toValidUri(String raw) {
        if (raw == null) return null;
        Components c = split(raw);

        StringBuilder out = new StringBuilder();
        if (c.scheme != null) out.append(c.scheme).append(':');
        if (c.authority != null) out.append("//").append(encode(c.authority, SAFE_AUTH, false));
        if (c.path != null) out.append(encode(c.path, SAFE_PATH, false));
        if (c.query != null) out.append('?').append(encode(c.query, SAFE_QUERY_FRAGMENT, false));
        if (c.fragment != null) out.append('#').append(encode(c.fragment, SAFE_QUERY_FRAGMENT, false));
        return out.toString();
    }

    public static boolean isValidIri(String value) {
        if (value == null || value.isBlank()) return false;
        IRI iri = IRI_FACTORY.create(value);
        return !iri.hasViolation(false);
    }

    public static boolean isValidUri(String value) {
        if (value == null || value.isBlank()) return false;
        // URI validity as ASCII-only policy
        String normalized = toValidUri(value);
        return value.equals(normalized) && hasScheme(value);
    }

    private static boolean hasScheme(String s) {
        int i = s.indexOf(':');
        if (i <= 0) return false;
        String scheme = s.substring(0, i);
        return scheme.matches("^[a-zA-Z][a-zA-Z0-9+.-]*$");
    }

    private static String encode(String s, String extraSafe, boolean allowUnicodeIriChars) {
        StringBuilder out = new StringBuilder();
        int i = 0;
        while (i < s.length()) {
            int cp = s.codePointAt(i);
            int charCount = Character.charCount(cp);

            // Preserve valid percent-encoded octets
            if (cp == '%' && i + 2 < s.length() && isHex(s.charAt(i + 1)) && isHex(s.charAt(i + 2))) {
                out.append('%').append(s.charAt(i + 1)).append(s.charAt(i + 2));
                i += 3;
                continue;
            }

            boolean safeAscii = cp < 128 && (Character.isLetterOrDigit(cp) || extraSafe.indexOf(cp) >= 0);
            boolean safeUnicode = allowUnicodeIriChars && isUcsChar(cp);

            if (safeAscii || safeUnicode) {
                out.appendCodePoint(cp);
            } else {
                byte[] bytes = new String(Character.toChars(cp)).getBytes(StandardCharsets.UTF_8);
                for (byte b : bytes) out.append('%').append(String.format("%02X", b & 0xFF));
            }

            i += charCount;
        }
        return out.toString();
    }

    private static boolean isHex(char c) {
        return (c >= '0' && c <= '9')
            || (c >= 'a' && c <= 'f')
            || (c >= 'A' && c <= 'F');
    }

    private static boolean isUcsChar(int cp) {
        return (cp >= 0xA0 && cp <= 0xD7FF)
            || (cp >= 0xF900 && cp <= 0xFDCF)
            || (cp >= 0xFDF0 && cp <= 0xFFEF)
            || (cp >= 0x10000 && cp <= 0xEFFFD);
    }

    private static Components split(String raw) {
        try {
            URI u = new URI(raw);
            if (u.isOpaque()) {
                // e.g. urn:uuid:...
                return new Components(u.getScheme(), null, u.getRawSchemeSpecificPart(), null, u.getRawFragment());
            }
            return new Components(u.getScheme(), u.getRawAuthority(), u.getRawPath(), u.getRawQuery(), u.getRawFragment());
        } catch (URISyntaxException e) {
            // Minimal fallback: still split reasonably to encode by component.
            return Components.fallback(raw);
        }
    }

    private static final class Components {
        final String scheme;
        final String authority;
        final String path;
        final String query;
        final String fragment;

        Components(String scheme, String authority, String path, String query, String fragment) {
            this.scheme = scheme;
            this.authority = authority;
            this.path = path;
            this.query = query;
            this.fragment = fragment;
        }

        static Components fallback(String raw) {
            String scheme = null, authority = null, path, query = null, fragment = null;
            int colon = raw.indexOf(':');
            String rest = raw;
            if (colon > 0) {
                scheme = raw.substring(0, colon);
                rest = raw.substring(colon + 1);
            }

            int hash = rest.indexOf('#');
            if (hash >= 0) {
                fragment = rest.substring(hash + 1);
                rest = rest.substring(0, hash);
            }

            int q = rest.indexOf('?');
            if (q >= 0) {
                query = rest.substring(q + 1);
                rest = rest.substring(0, q);
            }

            if (rest.startsWith("//")) {
                String tmp = rest.substring(2);
                int slash = tmp.indexOf('/');
                if (slash >= 0) {
                    authority = tmp.substring(0, slash);
                    path = tmp.substring(slash);
                } else {
                    authority = tmp;
                    path = "";
                }
            } else {
                path = rest;
            }

            return new Components(scheme, authority, path, query, fragment);
        }
    }
}
```

**Review by sjaakd — CHANGES_REQUESTED** [review]
If not required, remove the configuration parameter. If required, name it consistently and the default should possibly sanitised.  So,


1. Fix config key mismatch: loader reads dcat.iri.encodeInvalidChars, while profile files use dcat.iri.encodeInvalid; current profiles won’t activate the feature.
2. Avoid partial policy: current code always replaces whitespace with , even when flag is false; that’s surprising behavior split. Make policy explicit and consistent.
3. Prevent double-encoding in mixed inputs: when one illegal char triggers buildValidIri, existing %20 can become %2520. Preserve valid %HH sequences during encoding.
4. Prefer validity by default: for standards compliance, I’d default to always valid output (encode or fail policy), and keep a temporary compatibility mode only if needed for migration.

**Comment by PaulBoon** [comment]
@sjaakd Thanks for the feedback, recently I was working on another project, but now I will have a look at it.

File: src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java (hunk 1 above)
Comment (PaulBoon) [reply src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java]: Good point, I like 'sanitizer', not sure why I used 'fixer', so I wil change it

File: src/main/java/io/gdcc/spi/export/dcat3/mapping/ResourceMapper.java (hunk 5 above)
Comment (PaulBoon) [reply src/main/java/io/gdcc/spi/export/dcat3/mapping/ResourceMapper.java]: Agree, is much clearer, make it into a 'sanitize' function

File: src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java (hunk 3 above)
Comment (PaulBoon) [reply src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java]: Good point, I will move it

File: src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java (hunk 2 above)
Comment (PaulBoon) [reply src/main/java/io/gdcc/spi/export/dcat3/config/validate/IRIFixer.java]: @sjaakd Nice code, I will use it, with tiny changes, but it looks like it is better at preventing double encoding.
